ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs

```diff
@@ -822,7 +824,19 @@ withConnectionManager ConnectionManagerArguments {
               Right (mutableConnState@MutableConnState { connVar }, transition) ->
                 do traceWith tracer (TrConnectionTimeWait connId)
                    when (cmTimeWaitTimeout > 0) $
-                     unmask (threadDelay cmTimeWaitTimeout)
+                     let -- make sure we wait at least 'cmTimeWaitTimeout', we
+                         -- ignore all 'AsyncCancelled' exceptions.
+                         forceThreadDelay delay | delay <= 0 = pure ()
+                         forceThreadDelay delay = do
+                           t <- getMonotonicTime
+                           unmask (threadDelay delay)
+                             `catch` \e -> 
+                                case fromException e
+                                of Just (AsyncCancelled) -> do
+                                     t' <- getMonotonicTime
+                                     forceThreadDelay (delay - t' `diffTime` t)
+                                   _ -> throwIO e
+                     in forceThreadDelay cmTimeWaitTimeout
```

Comment:
Maybe move the definition elsewhere?

This function ignores `AsyncCancelled` exceptions but does not re-throws them; should it?

Reply:
Good question.  The answer is yes, it ignores them to make sure it will delay for the given time.  We `throwTo ... AsyncCancelled` and we don't care how much time it will take for the thread to continue.  The exception is when the connection manager shuts down, there we use `cancel`, which will wait until the thread terminates (and thus `WaitTime` passes).  We could just kill the threads with another exception and not wait for the the final transitions (or rethrow `AsyncCancelled`), but this would break tests, since the final transition is done after `WaitTime` passes.

Reply:
Got it! Thanks!